crates/biome_cli/src/reporter/mod.rs

```diff
@@ -0,0 +1,206 @@
+    }
+    pub fn warnings(&self) -> u32 {
+        self.warnings
+    }
+    pub fn skipped(&self) -> usize {
+        self.skipped
+    }
+    pub fn suggested_fixes_skipped(&self) -> u32 {
+        self.suggested_fixes_skipped
+    }
+    pub fn diagnostics_not_printed(&self) -> u32 {
+        self.diagnostics_not_printed
+    }
+}
+
+/// When using this trait, the type that implements this trait is the one that holds the read-only information to pass around
+pub trait Reporter {
+    /// Writes the summary using the underling visitor
+    fn write(&mut self, visitor: &mut dyn ReporterVisitor) -> io::Result<()>;
+}
+
+/// When using this trait, the type that implements this trait is the one that will **write** the data, ideally inside a buffer
+pub trait ReporterVisitor {
+    /// Writes the summary in the underling writer
+    fn report_summary(
+        &mut self,
+        traversal_mode: &Execution,
+        summary: &TraversalSummary,
+    ) -> io::Result<()> {
+        let _ = (summary, traversal_mode);
+        Ok(())
+    }
+
+    /// Writes a diagnostic
+    fn report_diagnostics(&mut self, payload: &DiagnosticsPayload) -> io::Result<()> {
+        let _ = payload;
+        Ok(())
+    }
+
+    /// Reports the fixes that were skipped
+    fn report_skipped_fixes(&mut self, execution: &Execution, skipped: usize) -> io::Result<()> {
+        let _ = (execution, skipped);
+        Ok(())
+    }
+
+    /// Reports diagnostics that were not printed
+    fn report_not_printed_diagnostics(
+        &mut self,
+        execution: &Execution,
+        not_printed: usize,
+    ) -> io::Result<()> {
+        let _ = (execution, not_printed);
+        Ok(())
+    }
+
+    /// Reports the total files that were checked
+    fn report_total(
+        &mut self,
+        execution: &Execution,
+        total: usize,
+        duration: Duration,
+    ) -> io::Result<()> {
+        let _ = (execution, total, duration);
+        Ok(())
+    }
+
+    /// Reports the number of files that were changed/modified
+    fn report_changed(&mut self, execution: &Execution, changed: usize) -> io::Result<()> {
+        let _ = (execution, changed);
+        Ok(())
+    }
+
+    /// Reports the number of files that were skipped
+    fn report_skipped(&mut self, execution: &Execution, skipped: usize) -> io::Result<()> {
+        let _ = (execution, skipped);
+        Ok(())
+    }
+
+    /// Reports the number of errors emitted during the traversal
+    fn report_errors(&mut self, execution: &Execution, errors: usize) -> io::Result<()> {
+        let _ = (execution, errors);
+        Ok(())
+    }
+
+    /// Reports the number of warnings emitted during the traversal
+    fn report_warnings(&mut self, execution: &Execution, warnings: usize) -> io::Result<()> {
+        let _ = (execution, warnings);
+        Ok(())
+    }
+}
+
+/// This function reports the result of a traversal
+pub(crate) fn report<R, V>(
+    reporter: &mut R,
+    reporter_visitor: &mut V,
```

Comment:
You could use `impl Trait`.
```suggestion
    reporter: &mut impl Reporter,
    reporter_visitor: &mut impl ReporterVisitor,
```

Reply:
We can't the compiler can't infer the type otherwise. Just tried.